[collector] Make sure scheduler is stopped once Stop returns

olivielpeau opened the pull request:
### What does this PR do?

Makes sure the scheduler is fully stopped once `Scheduler.Stop` returns. This fixes panics that could occur when the collector was stopped (generally during agent shutdown).

### Motivation

The scheduler submits checks to run to the check runner by sending to an unbuffered channel:
`Scheduler.checksPipe` (same channel as`Runner.pending` for the check runner).

When the check runner stops, it closes this channel, so at that point we have to make sure that all producers to the channel have stopped sending to the channel, otherwise we get a `panic: send on closed channel`.

The scheduler (and the collector as a whole) has 2 kinds of producers to this channel:
* one-time check schedulers (for long-running checks)
* the `jobQueue`s (for regular checks)

Unfortunately the scheduler was not stopping properly before. When calling `Scheduler.Stop`:
* the one-time check schedulers, if they hadn't scheduled their check yet, would still try to send to the channel
* the scheduler would send `stop` messages to the `jobQueues`, but it would never actually check that these producers exited correctly/in time, and those that were already blocked on the send to the channel would never listen to the `stop` message

This fixes both issues:
* one-time check schedulers, in the unlikely event they're still trying to enqueue their check when `Stop` is called, now stop trying
* the `jobQueue`s now exit correctly when stop is called, even if they're trying to enqueue checks to the channel

### Additional Notes

Added a couple of tests that cause panics without this fix.

I've removed `Scheduler.Reload` since I don't think it works right now and it's not used anywhere. There are a few things happening in `Scheduler.Stop` that are not reversible (the main thing is that once `jobQueues` are stopped, they can't currently be started up again). If we want `Reload` to work in the future we can spend some time and re-implement it.

olivielpeau committed:
[collector] Make sure scheduler is stopped once Stop returns

truthbk commented on pkg/collector/scheduler/scheduler.go:
```diff
@@ -176,25 +187,16 @@ func (s *Scheduler) Stop(timeout ...time.Duration) error {
-	return errors.New("Unable to perform reload")
-}
-
 // stopQueues shuts down the timers for each active queue
```
Can we note here that this is a blocking operation?

truthbk commented on pkg/collector/scheduler/scheduler.go:
```diff
@@ -166,6 +169,14 @@ func (s *Scheduler) Stop(timeout ...time.Duration) error {
+	// and wait for them to exit
+	close(s.cancelOneTime)
+	s.wgOneTime.Wait()
+	s.cancelOneTime = make(chan bool)
```
Why re-create the channel and workgroup?

olivielpeau replied:
that was done to support `scheduler.Reload`, but since I removed it we don't have to re-create these. Going to remove them 👍

truthbk commented on pkg/collector/scheduler/scheduler_test.go:
```diff
@@ -126,6 +132,23 @@ func TestStop(t *testing.T) {
+	// once the scheduler is stopped, it should be safe to close this channel. Otherwise, this should panic
+	close(s.checksPipe)
+	// sleep to make the runtime schedule the hanging producer goroutines, if there are any left
+	time.Sleep(time.Millisecond)
```
any way we could assert the hanging producer goroutines were indeed scheduled? Otherwise, why sleep at all? Safety?

olivielpeau replied:
I'm not aware of and haven't found any way to check that the go scheduler did schedule the other hanging goroutines.

Here the test wants to make sure that the other goroutines don't cause panics once the `s.checksPipe` channel is closed. Locally I've been able to consistently make the test cause panics with such a sleep. Whereas without a sleep the hanging goroutines often are never scheduled and the tests pass with the faulty code.

So overall this seems like a reasonable way to force the hanging goroutines to be scheduled

truthbk reviewed (commented):
Looks great overall, just a couple of questions, nit.

olivielpeau commented on pkg/collector/scheduler/scheduler.go:
```diff
@@ -125,7 +124,9 @@ func (s *Scheduler) Run() {
 	go func() {
 		log.Debug("Starting scheduler loop...")
 
+		s.mu.Lock()
```
I _think_ I did this on purpose cut can't remember why 🤔 

Anyway you're right, I'll move the locking/unlocking inside `startQueues`/`stopQueues`.

olivielpeau commented on pkg/collector/scheduler/scheduler.go:
```diff
@@ -207,10 +209,30 @@ func (s *Scheduler) startQueues() {
+		defer s.wgOneTime.Done()
+		select {
+		case s.checksPipe <- check:
+		case <-s.cancelOneTime:
```
you're completely right 👍

olivielpeau committed:
Address review comments

olivielpeau committed:
[collector] scheduler cleanup: remove timeout and buffered channels

* The timeout didn't really make sense, as long as the scheduler is
  implemented correctly there's no good reason to have a timeout (and
  having one can hide bugs)
* Removed buffered channels of the scheduler, they can hide bugs and aren't really
  needed here.

truthbk reviewed (commented):
Looks good to me 👍  
(just a nit on a comment)

olivielpeau commented on pkg/collector/scheduler/scheduler.go:
```diff
@@ -203,7 +201,11 @@ func (s *Scheduler) stopQueues() {
 }
 
 // startQueues loads the timer for each queue
+// Not blocking
```
Hmm yeah you're right, acquiring the lock is a blocking operation, and I meant that `startQueue` is not blocking. Let me change the comment

olivielpeau committed:
[collector] Update comments after review

truthbk reviewed (approved):
🚑
